nixos/modules/security/pam.nix

```diff
@@ -370,156 +370,262 @@ let
       # Samba stuff to the Samba module.  This requires that the PAM
       # module provides the right hooks.
       text = mkDefault
-        (''
-          # Account management.
-          account required pam_unix.so
-          ${optionalString use_ldap
-              "account sufficient ${pam_ldap}/lib/security/pam_ldap.so"}
-          ${optionalString (config.services.sssd.enable && cfg.sssdStrictAccess==false)
-              "account sufficient ${pkgs.sssd}/lib/security/pam_sss.so"}
-          ${optionalString (config.services.sssd.enable && cfg.sssdStrictAccess)
-              "account [default=bad success=ok user_unknown=ignore] ${pkgs.sssd}/lib/security/pam_sss.so"}
-          ${optionalString config.krb5.enable
-              "account sufficient ${pam_krb5}/lib/security/pam_krb5.so"}
-          ${optionalString cfg.googleOsLoginAccountVerification ''
+        (
+          ''
+            # Account management.
+            account required pam_unix.so
+          '' +
+          optionalString use_ldap
```

Comment:
`'' + optionalString use_ldap ''` instead of on separate lines

Reply:
nope doesn't look better :/
```
      text = mkDefault ( ''
        # Account management.
        account required pam_unix.so
        '' + optionalString use_ldap ''
          account sufficient ${pam_ldap}/lib/security/pam_ldap.so
        '' + optionalString (config.services.sssd.enable && cfg.sssdStrictAccess==false) ''
          account sufficient ${pkgs.sssd}/lib/security/pam_sss.so
        '' + optionalString (config.services.sssd.enable && cfg.sssdStrictAccess) ''
          account [default=bad success=ok user_unknown=ignore] ${pkgs.sssd}/lib/security/pam_sss.so
        '' + optionalString config.krb5.enable ''
          account sufficient ${pam_krb5}/lib/security/pam_krb5.so
        '' + optionalString cfg.googleOsLoginAccountVerification ''
          account [success=ok ignore=ignore default=die] ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so
          account [success=ok default=ignore] ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_admin.so
        '' + ''
```

![image](https://user-images.githubusercontent.com/56650223/141593699-007ff9bf-cb94-4acb-b1fb-291e8b11aa68.png)

i think it just isn't possible to do these well

Reply:
yours is maybe a *little* bit more readable than the old one

Reply:
Maybe using a list and `(concatStringsSep "\n (filter "" []))"` would make the code better?

Reply:
Not tested:

```diff
diff --git a/nixos/modules/security/pam.nix b/nixos/modules/security/pam.nix
index c180af75bb1..1f02c9867fb 100644
--- a/nixos/modules/security/pam.nix
+++ b/nixos/modules/security/pam.nix
@@ -1015,78 +1015,76 @@ in
     security.apparmor.includes."abstractions/pam" = let
       isEnabled = test: fold or false (map test (attrValues config.security.pam.services));
       in
-      lib.concatMapStringsSep "\n"
+      (lib.concatMapStringsSep "\n"
         (name: "r ${config.environment.etc."pam.d/${name}".source},")
-        (attrNames config.security.pam.services) +
-      ''
-      mr ${getLib pkgs.pam}/lib/security/pam_filter/*,
-      mr ${getLib pkgs.pam}/lib/security/pam_*.so,
-      r ${getLib pkgs.pam}/lib/security/,
-      '' +
-      optionalString use_ldap ''
-         mr ${pam_ldap}/lib/security/pam_ldap.so,
-      '' +
-      optionalString config.services.sssd.enable ''
-        mr ${pkgs.sssd}/lib/security/pam_sss.so,
-      '' +
-      optionalString config.krb5.enable ''
-        mr ${pam_krb5}/lib/security/pam_krb5.so,
-        mr ${pam_ccreds}/lib/security/pam_ccreds.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.googleOsLoginAccountVerification)) ''
-        mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
-        mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_admin.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.googleOsLoginAuthentication)) ''
-        mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
-      '' +
-      optionalString (config.security.pam.enableSSHAgentAuth
-                     && isEnabled (cfg: cfg.sshAgentAuth)) ''
-        mr ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.fprintAuth)) ''
-        mr ${pkgs.fprintd}/lib/security/pam_fprintd.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.u2fAuth)) ''
-        mr ${pkgs.pam_u2f}/lib/security/pam_u2f.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.usbAuth)) ''
-        mr ${pkgs.pam_usb}/lib/security/pam_usb.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.oathAuth)) ''
-        "mr ${pkgs.oathToolkit}/lib/security/pam_oath.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.yubicoAuth)) ''
-        mr ${pkgs.yubico-pam}/lib/security/pam_yubico.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.duoSecurity.enable)) ''
-        mr ${pkgs.duo-unix}/lib/security/pam_duo.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.otpwAuth)) ''
-        mr ${pkgs.otpw}/lib/security/pam_otpw.so,
-      '' +
-      optionalString config.security.pam.enableEcryptfs ''
-        mr ${pkgs.ecryptfs}/lib/security/pam_ecryptfs.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.pamMount)) ''
-        mr ${pkgs.pam_mount}/lib/security/pam_mount.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.enableGnomeKeyring)) ''
-        mr ${pkgs.gnome3.gnome-keyring}/lib/security/pam_gnome_keyring.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.startSession)) ''
-        mr ${pkgs.systemd}/lib/security/pam_systemd.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.enableAppArmor)
-                     && config.security.apparmor.enable) ''
-        mr ${pkgs.apparmor-pam}/lib/security/pam_apparmor.so,
-      '' +
-      optionalString (isEnabled (cfg: cfg.enableKwallet)) ''
-        mr ${pkgs.plasma5Packages.kwallet-pam}/lib/security/pam_kwallet5.so,
-      '' +
-      optionalString config.virtualisation.lxc.lxcfs.enable ''
-        mr ${pkgs.lxc}/lib/security/pam_cgfs.so
-      '';
+          (attrNames config.security.pam.services)) +
+      (lib.concatStringsSep "\n" (lib.remove "" [
+        ''
+          mr ${getLib pkgs.pam}/lib/security/pam_filter/*,
+          mr ${getLib pkgs.pam}/lib/security/pam_*.so,
+          r ${getLib pkgs.pam}/lib/security/,
+        ''
+        (optionalString use_ldap "mr ${pam_ldap}/lib/security/pam_ldap.so,")
+        (optionalString config.services.sssd.enable "mr ${pkgs.sssd}/lib/security/pam_sss.so,")
+        (optionalString config.krb5.enable ''
+          mr ${pam_krb5}/lib/security/pam_krb5.so,
+          mr ${pam_ccreds}/lib/security/pam_ccreds.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.googleOsLoginAccountVerification)) ''
+          mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
+          mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_admin.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.googleOsLoginAuthentication)) ''
+          mr ${pkgs.google-compute-engine-oslogin}/lib/pam_oslogin_login.so,
+        '')
+        (optionalString (config.security.pam.enableSSHAgentAuth
+                         && isEnabled (cfg: cfg.sshAgentAuth)) ''
+          mr ${pkgs.pam_ssh_agent_auth}/libexec/pam_ssh_agent_auth.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.fprintAuth)) ''
+          mr ${pkgs.fprintd}/lib/security/pam_fprintd.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.u2fAuth)) ''
+          mr ${pkgs.pam_u2f}/lib/security/pam_u2f.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.usbAuth)) ''
+          mr ${pkgs.pam_usb}/lib/security/pam_usb.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.oathAuth)) ''
+          "mr ${pkgs.oathToolkit}/lib/security/pam_oath.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.yubicoAuth)) ''
+          mr ${pkgs.yubico-pam}/lib/security/pam_yubico.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.duoSecurity.enable)) ''
+          mr ${pkgs.duo-unix}/lib/security/pam_duo.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.otpwAuth)) ''
+          mr ${pkgs.otpw}/lib/security/pam_otpw.so,
+        '')
+        (optionalString config.security.pam.enableEcryptfs ''
+          mr ${pkgs.ecryptfs}/lib/security/pam_ecryptfs.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.pamMount)) ''
+          mr ${pkgs.pam_mount}/lib/security/pam_mount.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.enableGnomeKeyring)) ''
+          mr ${pkgs.gnome3.gnome-keyring}/lib/security/pam_gnome_keyring.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.startSession)) ''
+          mr ${pkgs.systemd}/lib/security/pam_systemd.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.enableAppArmor)
+                         && config.security.apparmor.enable) ''
+          mr ${pkgs.apparmor-pam}/lib/security/pam_apparmor.so,
+        '')
+        (optionalString (isEnabled (cfg: cfg.enableKwallet)) ''
+          mr ${pkgs.plasma5Packages.kwallet-pam}/lib/security/pam_kwallet5.so,
+        '')
+        (optionalString config.virtualisation.lxc.lxcfs.enable ''
+          mr ${pkgs.lxc}/lib/security/pam_cgfs.so
+        '')
+      ]));
   };

 }
```

Reply:
its easy to test these with a flake
```
{
  inputs = {
    #nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    nixpkgs.url = "/home/artturin/nixgits/my-nixpkgs";
  };

  outputs = inputs@{ self, nixpkgs }: {

    nixosConfigurations.vm = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      specialArgs = { inherit inputs; };
      modules = [
        ({ pkgs, ... }: {
          documentation.enable = false;
          users.mutableUsers = false;
          users.users.root = {
            password = "root";
          };
          users.users.user = {
            password = "user";
            isNormalUser = true;
            extraGroups = [ "wheel" ];
          };
        })
      ];
    };
    # So that we can just run 'nix run' instead of
    # 'nix build ".#nixosConfigurations.vm.config.system.build.vm" && ./result/bin/run-nixos-vm'
    defaultPackage.x86_64-linux = self.nixosConfigurations.vm.config.system.build.vm;
    defaultApp.x86_64-linux = {
      type = "app";
      program = "${self.defaultPackage.x86_64-linux}/bin/run-nixos-vm";
    };
  };
}
```
just `nix build` and the files will be in `./result/system/etc/pam.d`

Reply:
Well, I posted more for the idea than being the actual solution. I don't know what you think if this is more readable or not (I quite like it).

Reply:
would be nice if optionalString didn't return a empty line but i dont think thats possible
```
  optionalString =
    # Condition
    cond:
    # String to return if condition is true
    string: if cond then string else "";
```

Reply:
You can always create something like:

```nix
maybeString = cond: string: if cond then string else null;
```

And do:
```
(concatStringsSep "\n (remove null [ ... ]))"
```

Instead